### DISCUSSION TITLE
Help with unstable simulation using mesh refinement

### DISCUSSION THREAD

**Discussion by Cosimo89** [BLAST-WarpX/warpx#5505] [Q&A]
Dear WarpX community,

I am writing here because of a problem I am facing in what I thought would be a simple setup of laser-plasma interaction study.
I want to investigate the heating and interaction between electrons and layers of plasma with different material. The laser pulse regime is in the low-energy, long-duration, so not the typical application range for ultra-intense laser-plasma interactions.

I started by modifying the input script from the laser-ion acceleration case (https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Physics_applications/laser_ion/inputs_test_2d_laser_ion_acc)

In my first setup, I wanted to have a slab of Carbon at a given density (say $\sim30n_c$) and a low intensity laser pulse impinging on it. On a later stage I wanted to add an under-dense hydrogen ($n_c\leq 1$) after the Carbon slab. Ideally, I wanted to have ionization and coulomb binary collisions included.  I also want to use mesh refinement around the Carbon slab, to reduce the computational burden.

Below a sketch of the simulation setup:

![image](https://github.com/user-attachments/assets/c9549ba9-a0c8-46e9-a6d4-e8bd54d7669c)

The problem that I am facing is that in all investigated configurations, the simulation is unstable as non-physical electric fields seems to be generated at the interface between the refined and the coarse regions which lead to an abrupt explosion of the plasma (see images below). In all my trials, the problem seems to arise around the mesh refinement region, but it also appeared in a low-resolution case I performed without using mesh refinement.

Please note that this test is performed using on purpose a grid resolution of about 4 times coarser than a proper one for production. Can it be a resolution issue?

![image](https://github.com/user-attachments/assets/b1a36e74-c4d5-4d4a-a1c2-b2b4ce4377e3)
![image](https://github.com/user-attachments/assets/3d61de1e-d286-45f9-89e6-caec084d004b)
![image](https://github.com/user-attachments/assets/284d7b68-d413-42a3-9319-63b4b562104e)

I performed many different trials to debug my input script (i.e. different boundary conditions, domain sizes, material, no ionization), but all simulations showed similar behavior.
I have the feeling I am committing some bad setup error, but I am afraid not be expert enough to spot it, so I invoke the help of the experts!

Any help/input/ feedback on the input file (attached below) is highly appreciated.   
  
   
[inputs_test_2d.txt](https://github.com/user-attachments/files/18077089/inputs_test_2d.txt)

Thank you in advance and best regards,
Cosimo

**Comment by n01r** [comment]
Hi @Cosimo89, thank you for the detailed description!

I have seen something similar where my simulation became unstable due to low resolution where with the choice of electromagnetic solver a lower value for the `warpx.cfl` would have made it more stable. But in my case, the patches did not cut through the target. Perhaps this could be the main issue here.

Could you try to include the whole target in the refined patch? I am wondering if the damped motion of heated electrons would cause fields of different frequencies some of which cannot be properly resolved. 
@EZoni, @oshapoval maybe you could weigh in on that? 

Side note:
This line in the input script you supplied looks a bit odd. 
```
laser1.a0           = 0 .01              # maximum amplitude of the laser field [V/m]
```
There is a space between the zero and the `.01`. But that should be unrelated.

**Comment by Cosimo89** [comment]
Hi @n01r ,

thank you for the reply, really appreciate it!

I actually did a large number of tests, varying materials, refinement patch, active models and so on. 
Unfortunately I don't have what you ask for the simple Carbon slab case, but I have it when I was switched to Lithium. Below some screenshots:

### Case A: Material switched to Li, same resolution with refined patch applied to the whole slab, particles-per-cell = 10x10.  
 
Snapshot time: 110fs
Un-physical electric field appears at the edges of the slab (top and bottom) and simulation becomes unstable.  

![image](https://github.com/user-attachments/assets/9c8bc216-9ee9-4145-b157-894eeb49d8ca)

### Case B:  Material switched to Li, roughly 6x larger base resolution with refined patch applied to the whole slab, particles-per-cell = 5x5.  

Snapshot time: 51fs
Similar story, un-physical electric field appears at the edges and simulation is unstable
 
![image](https://github.com/user-attachments/assets/c7b8f664-3312-4599-9320-46f48d550d41).  
  
To summarize: every time I use a refined region where the plasma reaches the edges of the refinement, the simulation becomes unstable. This effect occurs in both low and medium resolution cases (a fully resolved case would require 5x more resolution than case B), and in both cases where the patch cuts trough the plasma or simply is in contact with the domain boundaries.

PS: about the space after the definition of a0, everything looks good, at least from the ```warpx_used_inputs``` file:

```
laser1.a0 = 0 .01

```

But thanks for pointing that out!

**Reply by n01r** [reply]
Hi @Cosimo89, 

I see. 
There might be a few more things you can try: 
*  if you have wiggle room with your resources, you could extend the box more in the transverse direction
* you could consider making the transverse direction have periodic boundaries (given it's far enough from the interesting region)
* You could try to let your slab target end early before the edge of the box, and possibly apply a soft (Gaussian) ramp so that the plasma edge would not be super sharp (which should prevent strong fields)

I would stay with the same material for now and just try to get the fields under control.


Your interaction time is very long, but nevertheless, these fields at the edges don't look great.
In relation to your plots above: at which time do the fields at the edges appear?

**Reply by n01r** [reply]
Hi @Cosimo89, would you also be able to do a run without mesh refinement at all?

I brought this up in our developer meeting today and this was a point that came up and that makes sense to check.
It is not unheard of that when the resolution isn't great and depending on the algorithms used, one might already see boundary effects like that.

Also, another colleague reported that when the edge of the fine patch is basically on the boundary of the simulation box but not exactly, they ran into artifacts, too. One way to fix that is to have the fine patch either extend slightly beyond the box, or have it stay clear of the global boundaries and their PML buffer regions, as suggested above (but again, the plasma target should not be intersected).

**Comment by Cosimo89** [comment]
Hi @n01r 

first of all, thank you a lot for the interest you are putting into this. 
#### Reply to comment 1:
About the suggestions in your first comment: they are all interesting. In relation to applying periodic boundaries to the transverse direction, I thought about it, but then I assumed that the open boundaries would be the most stable. I will also apply the other suggestions.

In relation to the simulation time of the snapshot posted previously, I edited the comment and added it. As you can see the instability happens very early (110 fs and 51 fs, respectively).  

#### Reply to comment 2:

I performed some simulations of the slab without refinement. They were stable, but way too slow to have any chance of completion.   
What I am doing now is switching to a spherical target. In this way I manage to stay away from the boundaries, and since there is much less plasma I am avoiding refinement. Simulations are very stable, I managed to reach 10kfs and now even restarting from there.

Very interesting also is the point about the refined patch matching the boundaries. I will definitely try what you suggest there if in the future I will use refinement patch again.

If you want me to run some specific case associated to the slab to try to fix the problem, feel free to ask, I will try to run something when I find the space in the cluster :)

**Reply by n01r** [reply]
Thank you for editing in the times!

I see that the artifacts appear way earlier than before. That may be connected to what I said in _comment 2_ where there may be a weird overlap effect between the boundary PM buffers and the refinement patch PML buffers. For foil targets that reach out far beyond your simulated box, it is usually important to make them transversely large enough. That is to prevent that within the time of most interesting interactions, significant amounts of energy or particles reach the boundary that cause artifacts that travel back and influence your interaction. True `thermal` particle boundaries, emulating that the medium continues beyond the box, are difficult. But your setup is also tricky because it runs for so long. Maybe you are studying preplasma formation, for which fluid codes are usually used. But I get it because they mostly lack ionization effects and fail to represent kinetic effects that set on during the 1e16 to 1e17 W/cm² intensity range.

**Reply by Cosimo89** [reply]
Thank you @n01r for all the useful explanations and tip, it really helps.

Indeed I am interested in studying the electrons dynamics and plasma heating due to hot electrons. My final aim was to simulate a spherical target, and I thought that starting from the foil was simpler. Actually it seems that the spherical case is easier, probably, as you hint, because the plasma is fully inside the simulation box and it does not touch the boundaries.

Again, thanks for taking the time to reply to my question, I learned a lot!
